File: go/apskit/runtime.go
package apskit

import "fmt"

// ExecutionsFor returns the (scenarioIndex, exampleIndex) pairs the runtime
// must execute. A scenario without examples yields one pair with example 0.
func ExecutionsFor(f *Feature) [][2]int {
	var out [][2]int
	for sIdx, s := range f.Scenarios {
		if len(s.Examples) == 0 {
			out = append(out, [2]int{sIdx, 0})
			continue
		}
		for eIdx := range s.Examples {
			out = append(out, [2]int{sIdx, eIdx})
		}
	}
	return out
}

func exampleFor(s Scenario, exampleIndex int) (Example, error) {
	if len(s.Examples) == 0 {
		if exampleIndex != 0 {
			return nil, fmt.Errorf("scenario %q has no examples; only exampleIndex=0 is valid", s.Name)
		}
		return Example{}, nil
	}
	if exampleIndex < 0 || exampleIndex >= len(s.Examples) {
		return nil, fmt.Errorf("example index %d out of range for scenario %q", exampleIndex, s.Name)
	}
	out := make(Example, len(s.Examples[exampleIndex]))
	for k, v := range s.Examples[exampleIndex] {
		out[k] = v
	}
	return out, nil
}

func runStep(step Step, world World, example Example, reg *Registry) error {
	// Resolve the handler first so an unsupported step gives the most useful
	// error before we check anything else; then validate required example
	// values; then invoke. Same order in all four language runtimes.
	fn, err := reg.Resolve(step.Text)
	if err != nil {
		return err
	}
	for _, p := range step.Parameters {
		if _, ok := example[p]; !ok {
			return fmt.Errorf("step %q references missing example value %q", step.Text, p)
		}
	}
	return fn(world, example)
}

// RunExecution loads the IR file and runs one scenario/example execution. If
// reg is nil, DefaultRegistry is used.
func RunExecution(irPath string, scenarioIndex, exampleIndex int, reg *Registry) error {
	if reg == nil {
		reg = DefaultRegistry
	}
	feature, err := LoadIR(irPath)
	if err != nil {
		return err
	}
	if scenarioIndex < 0 || scenarioIndex >= len(feature.Scenarios) {
		return fmt.Errorf("scenario index %d out of range; feature has %d scenarios", scenarioIndex, len(feature.Scenarios))
	}
	scenario := feature.Scenarios[scenarioIndex]
	example, err := exampleFor(scenario, exampleIndex)
	if err != nil {
		return err
	}
	world := World{}
	for _, step := range feature.Background {
		if err := runStep(step, world, example, reg); err != nil {
			return err
		}
	}
	for _, step := range scenario.Steps {
		if err := runStep(step, world, example, reg); err != nil {
			return err
		}
	}
	return nil
}

File: rust/aps-kit/src/runtime.rs
use std::path::Path;

use crate::ir::{load_ir, Feature, Scenario, Step};
use crate::registry::{default_registry, Example, Registry, StepError, World};

/// Yields (scenarioIndex, exampleIndex) pairs the runtime must execute.
pub fn executions_for(feature: &Feature) -> Vec<(usize, usize)> {
    let mut out = Vec::new();
    for (s_idx, scenario) in feature.scenarios.iter().enumerate() {
        if scenario.examples.is_empty() {
            out.push((s_idx, 0));
        } else {
            for e_idx in 0..scenario.examples.len() {
                out.push((s_idx, e_idx));
            }
        }
    }
    out
}

fn example_for(scenario: &Scenario, example_index: usize) -> Result<Example, StepError> {
    if scenario.examples.is_empty() {
        if example_index != 0 {
            return Err(StepError::Failure(format!(
                "scenario {:?} has no examples; only exampleIndex=0 is valid",
                scenario.name
            )));
        }
        return Ok(Example::new());
    }
    let row = scenario
        .examples
        .get(example_index)
        .ok_or_else(|| StepError::Failure(format!(
            "example index {} out of range for scenario {:?}",
            example_index, scenario.name
        )))?;
    Ok(row.clone())
}

fn run_step(
    step: &Step,
    world: &mut World,
    example: &Example,
    registry: &Registry,
) -> Result<(), StepError> {
    for param in &step.parameters {
        if !example.contains_key(param) {
            return Err(StepError::MissingParameter {
                step: step.text.clone(),
                param: param.clone(),
            });
        }
    }
    registry.invoke(&step.text, world, example)
}

/// Runs one (scenario, example) execution from the IR file at `ir_path`.
/// Uses `default_registry()` when `registry` is `None`.
pub fn run_execution<P: AsRef<Path>>(
    ir_path: P,
    scenario_index: usize,
    example_index: usize,
    registry: Option<&Registry>,
) -> Result<(), StepError> {
    let feature = load_ir(&ir_path).map_err(StepError::Failure)?;
    let scenario = feature
        .scenarios
        .get(scenario_index)
        .ok_or_else(|| StepError::Failure(format!(
            "scenario index {scenario_index} out of range; feature has {} scenarios",
            feature.scenarios.len()
        )))?;
    let example = example_for(scenario, example_index)?;
    let reg = registry.unwrap_or_else(|| default_registry());
    let mut world: World = serde_json::json!({});
    for step in &feature.background {
        run_step(step, &mut world, &example, reg)?;
    }
    for step in &scenario.steps {
        run_step(step, &mut world, &example, reg)?;
    }
    Ok(())
}
